cloud/openstack/os_security_group_rule.py

```diff
@@ -0,0 +1,257 @@
+      description:
+        - Starting port
+      required: true
+   port_range_max:
+      description:
+        - Ending port
+      required: true
+   remote_ip_prefix:
+      description:
+        - Source IP address(es) in CIDR notation (exclusive with remote_group)
+      required: false
+   remote_group:
+      description:
+        - ID of Security group to link (exclusive with remote_ip_prefix)
+      required: false
+   ethertype:
+      description:
+        - Must be IPv4 or IPv6, and addresses represented in CIDR must
+          match the ingress or egress rules. Not all providers support IPv6.
+      choices: ['IPv4', 'IPv6']
+      default: IPv4
+   direction:
+     description:
+        - The direction in which the security group rule is applied. Not
+          all providers support egress.
+     choices: ['egress', 'ingress']
+     default: ingress
+   state:
+     description:
+       - Should the resource be present or absent.
+     choices: [present, absent]
+     default: present
+requirements: ["shade"]
+'''
+# TODO(mordred): add ethertype and direction
+
+EXAMPLES = '''
+# Create a security group rule
+- os_security_group_rule:
+    cloud: mordred
+    security_group: foo
+    protocol: tcp
+    port_range_min: 80
+    port_range_max: 80
+    remote_ip_prefix: 0.0.0.0/0
+'''
+
+RETURN = '''
+id:
+  description: Unique rule UUID.
+  type: string
+direction:
+  description: The direction in which the security group rule is applied.
+  type: string
+  sample: 'egress'
+ethertype:
+  description: One of IPv4 or IPv6.
+  type: string
+  sample: 'IPv4'
+port_range_min:
+  description: The minimum port number in the range that is matched by
+               the security group rule.
+  type: int
+  sample: 8000
+port_range_max:
+  description: The maximum port number in the range that is matched by
+               the security group rule.
+  type: int
+  sample: 8000
+protocol:
+  description: The protocol that is matched by the security group rule.
+  type: string
+  sample: 'tcp'
+remote_ip_prefix:
+  description: The remote IP prefix to be associated with this security group rule.
+  type: string
+  sample: '0.0.0.0/0'
+security_group_id:
+  description: The security group ID to associate with this security group rule.
+  type: string
+'''
+
+
+def _find_matching_rule(module, secgroup):
+    """
+    Find a rule in the group that matches the module parameters.
+    :returns: The matching rule dict, or None if no matches.
+    """
+    protocol = module.params['protocol']
+    port_range_min = module.params['port_range_min']
+    port_range_max = module.params['port_range_max']
+    remote_ip_prefix = module.params['remote_ip_prefix']
+    ethertype = module.params['ethertype']
+    direction = module.params['direction']
+
+    for rule in secgroup['security_group_rules']:
+        # No port, or -1, will be returned from shade as None
+        if rule['port_range_min'] is None:
+            rule_port_range_min = -1
```

Comment:
I believe you need a shade patch I created for this. Will verify tomorrow. 

Sent from my iPhone

> On Jun 30, 2015, at 6:13 PM, Jesse Keating <notifications@github.com> wrote:
> 
> In cloud/openstack/os_security_group_rule.py:
> 
> > +def _find_matching_rule(module, secgroup):
> > +    """
> > +    Find a rule in the group that matches the module parameters.
> > +    :returns: The matching rule dict, or None if no matches.
> > +    """
> > +    protocol = module.params['protocol']
> > +    port_range_min = module.params['port_range_min']
> > +    port_range_max = module.params['port_range_max']
> > +    remote_ip_prefix = module.params['remote_ip_prefix']
> > +    ethertype = module.params['ethertype']
> > +    direction = module.params['direction']
> > +
> > +    for rule in secgroup['security_group_rules']:
> > +        # No port, or -1, will be returned from shade as None
> > +        if rule['port_range_min'] is None:
> > +            rule_port_range_min = -1
> This is breaking comparison on rules that do not have port ranges. Rules without port ranges are coming back from shade as None and this code is setting it to -1, however rules being passed to the module are also coming through as None and not being converted here. Can we do away with the conversion to -1, or match on both -1 AND None?
> 
> —
> Reply to this email directly or view it on GitHub.
>